internal/oauthserver/oauth_server.go

```diff
@@ -0,0 +1,289 @@
+func NewOAuthServer(
+	oauthClient auth.OAuthClient,
+	sessionStore sessions.Store,
+	directory identity.Directory,
+) *OAuthServer {
+	storage := newStore()
+	config := &fosite.Config{
+		GlobalSecret:               []byte("my super secret signing password"),
+		SendDebugMessagesToClients: true,
+	}
+	provider := compose.Compose(
+		config,
+		storage,
+		compose.NewOAuth2HMACStrategy(config),
+		compose.OAuth2AuthorizeExplicitFactory,
+		compose.OAuth2RefreshTokenGrantFactory,
+		compose.OAuth2PKCEFactory,
+	)
+	// Register types for session serialization
+	gob.Register(&authRequestFlash{})
+	gob.Register(auth.AuthorizeState{})
+	return &OAuthServer{
+		storage:      storage,
+		provider:     provider,
+		oauthClient:  oauthClient,
+		sessionStore: sessionStore,
+		directory:    directory,
+	}
+}
+
+// HandleAuthorize processes OAuth 2.0 authorization requests from the client.
+//
+// This handler initiates the authorization flow by:
+//  1. Validating the client's authorize request
+//  2. Resolving the user's atproto handle
+//  3. Initiating authorization with the user's PDS
+//  4. Storing the request context in a cookie session
+//  5. Redirecting to the PDS for user authentication
+//
+// The request must include a "handle" form parameter with the user's handle
+// (e.g., "alice.bsky.social").
+//
+// Returns an error if the authorization request is invalid or if any step in the
+// authorization process fails. OAuth-specific errors are written directly to the
+// response using the OAuth error format.
+func (o *OAuthServer) HandleAuthorize(
+	w http.ResponseWriter,
+	r *http.Request,
+) {
+	ctx := r.Context()
+	requester, err := o.provider.NewAuthorizeRequest(ctx, r)
+	if err != nil {
+		o.provider.WriteAuthorizeError(ctx, w, requester, err)
+		return
+	}
+	if r.ParseForm() != nil {
+		utils.LogAndHTTPError(w, err, "failed to parse form", http.StatusBadRequest)
+		return
+	}
+	handle := r.Form.Get("handle")
+	atid, err := syntax.ParseAtIdentifier(handle)
+	if err != nil {
+		utils.LogAndHTTPError(w, err, "failed to parse handle", http.StatusBadRequest)
+		return
+	}
+	id, err := o.directory.Lookup(ctx, *atid)
+	if err != nil {
+		utils.LogAndHTTPError(w, err, "failed to lookup identity", http.StatusInternalServerError)
+		return
+	}
+	dpopKey, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		utils.LogAndHTTPError(w, err, "failed to generate key", http.StatusInternalServerError)
+		return
+	}
+	dpopClient := auth.NewDpopHttpClient(dpopKey, &nonceProvider{})
+	redirect, state, err := o.oauthClient.Authorize(dpopClient, id)
+	if err != nil {
+		utils.LogAndHTTPError(
+			w,
+			err,
+			"failed to initiate authorization",
+			http.StatusInternalServerError,
+		)
+		return
+	}
+	authorizeSession, _ := o.sessionStore.New(r, sessionName)
+	authorizeSession.AddFlash(&authRequestFlash{
+		ClientID:       requester.GetClient().GetID(),
+		RedirectURI:    requester.GetRedirectURI(),
```

Comment:
this is the oauth server's /callback endpoint right? or no ..

Reply:
no this is the client's redirect uri which we eventually want to redirect back to. First we redirect to atproto which will redirect to the oauth server's `/callback` and then we redirect back to this one with the authorization code